deeptime/markov/msm/tram/_tram_dataset.py

```diff
@@ -0,0 +1,494 @@
+    """
+
+    def __init__(self, dtrajs, bias_matrices, ttrajs=None, n_therm_states=None, n_markov_states=None, lagtime=1,
+                 count_mode='sliding'):
+        self.lagtime = lagtime
+        self.count_mode = count_mode
+        self.dtrajs = dtrajs
+        self.ttrajs = ttrajs
+        self.bias_matrices = bias_matrices
+        self.count_models = []
+
+        self._ensure_correct_data_types()
+
+        if n_therm_states is None:
+            self._n_therm_states = _determine_n_therm_states(self.dtrajs, self.ttrajs)
+        else:
+            self._n_therm_states = n_therm_states
+
+        if n_markov_states is None:
+            self._n_markov_states = _determine_n_states(self.dtrajs)
+        else:
+            self._n_markov_states = n_markov_states
+
+        # validate all dimensions
+        self._check_dimensions()
+        # compute the count models
+        self._compute_counts()
+
+    #: All possible connectivity modes
+    connectivity_options = ["post_hoc_RE", "BAR_variance", "summed_count_matrix", None]
+
+    @property
+    def tram_input(self):
+        r""" The TRAMInput object containing the data needed for estimation. """
+        return tram.TRAMInput(self.state_counts, self.transition_counts, self.dtrajs, self.bias_matrices)
+
+    @property
+    def n_therm_states(self):
+        r""" The number of thermodynamic states. """
+        return self._n_therm_states
+
+    @property
+    def n_markov_states(self):
+        r""" The number of Markov states. """
+        return self._n_markov_states
+
+    @property
+    def transition_counts(self):
+        r""" ndarray(n, m, m)
+        The transition counts matrices. transition_counts[k] contains the transition counts for thermodynamic state
+        k, based on the TransitionCountModel of state k. transition_counts[k][i][j] equals the number of observed
+        transitions from Markov state :math:`i` to Markov state :math:`j`, in thermodynamic state :math:`k`.
+
+        The transition counts for every thermodynamic state are shaped to contain all possible markov states, even the
+        ones without counts in that thermodynamic state. This is done so that the _tram_bindings receives count matrices
+        that are all the same shape, which is easier to handle (matrices are padded with zeros for all empty states that
+        got  dropped by the TransitionCountModels).
+        """
+        transition_counts = np.zeros((self.n_therm_states, self.n_markov_states, self.n_markov_states), dtype=np.intc)
+
+        for k in range(self.n_therm_states):
+            model_k = self.count_models[k]
+            if model_k.count_matrix.sum() > 0:
+                i_s, j_s = np.meshgrid(model_k.state_symbols, model_k.state_symbols)
+                # place submodel counts in our full-sized count matrices
+                transition_counts[k, i_s, j_s] = model_k.count_matrix.T
+
+        return transition_counts
+
+    @property
+    def state_counts(self):
```

Comment:
Same as `transition_counts`: You can document the return type with `:type:` and `:getter:`; the result can be cached by using `cached_property`